Correctly check for ragged array

pmrv opened the pull request:
Fixes #502 in my testing, but haven't tested yet if something else depended on this weird behavior.

pmrv committed:
Correctly check for ragged arrays

Caused a performance regression, by writting arrays as dtype=object

pmrv committed:
Add test for _is_ragged_array

jan-janssen commented:
@pmrv Great, can you maybe also add a quick performance test? Just so we do not forget about it.

niklassiemer commented:
Thanks, indeed! LGTM! However, I am wondering why the compatibility workflows (already) ran?! 

Edit: Made a new issue for this #504

pmrv commented:
Looking at the diff again, I see that the checks are not actually equivalent.  The original code checks whether the shapes of arrays in `value` are all equal except the first dimension.  It still sounds to me like that check wants to check whether `value` is ragged only in the first dimension, but it's still a bit weird to me.

liamhuber commented on pyiron_base/generic/hdfio.py:
```diff
@@ -27,6 +28,19 @@
+        bool: True if elements of value are not all of the same shape
+    """
+    shape_lst = [np.shape(sub) for sub in value]
+    return len(set(shape_lst)) > 1
```
I was curious what was faster: `set`, `numpy.unique`, or `pandas.unique`. Operating on `[np.random.randint(50) for _ in range(2000)]` it was `set` *by an order of magnitude!*

Anyhow, no change here, I just thought this was interesting.

liamhuber commented on pyiron_base/generic/hdfio.py:
```diff
@@ -165,8 +179,7 @@ def __setitem__(self, key, value):
         ):
-            shape_lst = [np.shape(sub) for sub in value]
-            if all([shape_lst[0][1:] == t[1:] for t in shape_lst]):
+            if _is_ragged_array(value):
```
I appreciate this refactor. Makes things very readable.

liamhuber reviewed (approved):
lgtm. I also agree the check has changed, but actually I don't understand exactly what the previous check was trying to accomplish and the current check is very clear so...go for it?

liamhuber commented:
@raynol-dsouza keep an eye on this -- I guess it will probably be merged quickly enough to give you some QoL improvement for the calculations we talked about yesterday!

pmrv committed:
Add benchmark test

pmrv commented:
I've copied the benchmark workflow over from `pyiron_atomistics`.

pmrv commented:
Whoops, turns out the default number of repetitions for `timeit` is a tiny bit crazy. Using a lower number now.

niklassiemer reviewed (approved):
LGTM!

niklassiemer commented:
It seems like, we actually depended on a bug in h5io to return float arrays instead of object arrays to allow this bug to work! (see https://github.com/pyiron/pyiron_atomistics/issues/409)

pmrv committed:
Don't check shapes on regular numpy arrays

This makes the common path (ie. value = ndarray) *much* faster for large arrays
